### PR TITLE
Separate doublebuffer

### PR DESCRIPTION
**Pull request by emilwestergren**
All logic for double buffering is now separated in a specialized implementation of `VideoProcessor`.

@sebastianbaginski peer review

### CODE REVIEW

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+ */
+
+#include "DoubleBufferVideoProcessor.h"
+
+namespace imint {
+std::map<buffer_handle_t*, double_buffer_t> doubleBuffers;
+
+DoubleBufferVideoProcessor::DoubleBufferVideoProcessor(const char* cameraId)
+		: VideoProcessor(cameraId) {
+	pthread_mutex_init(&this->doubleBufferMutex, NULL);
```
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: Why not `std::mutex` ? I think it's a lot simpler to use.

Comment (emilwestergren) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: Yes, I will change this later. Not in this PR though.

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+}
+DoubleBufferVideoProcessor::~DoubleBufferVideoProcessor() {
+	pthread_mutex_destroy(&this->doubleBufferMutex);
+	doubleBuffers.clear();
```
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: With `std::mutex` you could have empty destructor body. `doubleBuffers.clear()` is redundant.

File: HAL/HAL3/DoubleBufferVideoProcessor.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+#ifndef HAL_HAL3_DOUBLEBUFFERVIDEOPROCESSOR_H_
+#define HAL_HAL3_DOUBLEBUFFERVIDEOPROCESSOR_H_
+#include "VideoProcessor.h"
+
+namespace imint {
+
+typedef struct double_buffer {
+	sp<GraphicBuffer> cameraBuffer;
+	camera3_stream_buffer_t *frameworkBuffer;
```
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.h]: `camera3_stream_buffer_t *frameworkBuffer = nullptr;` ?

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+void DoubleBufferVideoProcessor::removeDoubleBuffer(buffer_handle_t* handle) {
+	double_buffer_t result = getDoubleBuffer(handle);
+	pthread_mutex_lock(&this->doubleBufferMutex);
+	doubleBuffers.erase(handle);
+	delete(result.frameworkBuffer);
+	pthread_mutex_unlock(&this->doubleBufferMutex);
+	result.cameraBuffer = NULL;
+}
+
+int DoubleBufferVideoProcessor::registerStreamBuffers(const camera3_stream_buffer_set_t *buffer_set) {
```
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: I may be wrong, but I think I saw this code somewhere before

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp (hunk 2 above)
Comment (emilwestergren) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: Why is `doubleBuffers.clear()` redundant? Note that it is a global variable (which it probably shouldn't be but anyway).

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+	pthread_mutex_lock(&this->doubleBufferMutex);
+	result = doubleBuffers[handle];
+	pthread_mutex_unlock(&this->doubleBufferMutex);
+	return result;
+}
+void DoubleBufferVideoProcessor::removeDoubleBuffer(buffer_handle_t* handle) {
+	double_buffer_t result = getDoubleBuffer(handle);
+	pthread_mutex_lock(&this->doubleBufferMutex);
+	doubleBuffers.erase(handle);
+	delete(result.frameworkBuffer);
```
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: "You don't need this brackets" as we use to say in the other repo :P

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp (hunk 2 above)
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: A global... that explains it.

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+	pthread_mutex_unlock(&this->doubleBufferMutex);
+	result.cameraBuffer = NULL;
```
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: This is not needed.

**Comment by emilwestergren** [comment]
I'll fix most of your comments in a later PR if that's ok, this one is mostly about moving code around so I don't know if I want all these improvements in the same PR.

**Comment by emilwestergren** [comment]
Or nvm, I'll just fix it right away.

**Comment by sebastianbaginski** [comment]
OK, your call.

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp (hunk 5 above)
Comment (emilwestergren) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: What do you mean?

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp (hunk 4 above)
Comment (emilwestergren) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: I think you're wrong, this function has been moved from `VideoProcessor` to `DoubleBufferVideoProcessor` (great name right?)

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp (hunk 5 above)
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: `delete result.frameworkBuffer;`

File: HAL/HAL3/DoubleBufferVideoProcessor.cpp (hunk 4 above)
Comment (sebastianbaginski) [inline HAL/HAL3/DoubleBufferVideoProcessor.cpp]: Still better than `IGraphBuilder2`

**Pull request merged**